core/src/client/async_client/mod.rs

```diff
@@ -726,95 +771,157 @@ async fn background_task<S, R>(
-				)
-				.await;
-
-				// Advance frontend, save backend.
-				message_fut = future::select(frontend.next(), backend);
-				exit_or_ping_fut = exit_or_ping;
-			}
-			// Message received from the backend.
-			Either::Left((Either::Right((backend_value, frontend)), exit_or_ping)) => {
-				if let Err(err) = handle_backend_messages::<S, R>(
-					backend_value,
-					&mut manager,
-					&mut sender,
-					max_buffer_capacity_per_subscription,
-				)
-				.await
+				if let Err(e) =
+					handle_frontend_messages(msg, &manager, &mut sender, max_buffer_capacity_per_subscription).await
 				{
-					tracing::error!("[backend]: {}", err);
-					let _ = front_error.send(err);
-					break;
+					tracing::error!("Could not send message: {e}");
+					break Err(Error::Transport(e.into()));
 				}
-				// Advance backend, save frontend.
-				message_fut = future::select(frontend, backend_event.next());
+
 				exit_or_ping_fut = exit_or_ping;
+				rx_item = frontend.next();
 			}
-			// The client is closed.
-			Either::Right((Either::Left((_, _)), _)) => {
-				break;
-			}
-			// Submit ping interval was triggered if enabled.
-			Either::Right((Either::Right((_, on_exit)), msg)) => {
+
+			// Ping
+			Either::Right((Either::Right((_, on_exit)), rx)) => {
 				if let Err(err) = sender.send_ping().await {
-					tracing::error!("[backend]: Could not send ping frame: {}", err);
-					let _ = front_error.send(Error::Custom("Could not send ping frame".into()));
-					break;
+					tracing::error!("[backend]: Could not send ping frame: {err}");
+					break Err(Error::Custom("Could not send ping frame".into()));
 				}
-				message_fut = msg;
 				exit_or_ping_fut = future::select(on_exit, ping_fut(ping_interval));
+				rx_item = rx;
 			}
-		};
-	}
 
-	// Wake the `on_disconnect` method.
+			// Closed
+			Either::Right((Either::Left(_), _)) => {
+				break Ok(());
+			}
+		}
+	};
+
 	frontend.close();
-	// Send close message to the server.
-	_ = sender.close().await;
+	let _ = sender.close().await;
+	let _ = close_tx.send(res).await;
 }
 
-fn unparse_error(raw: &[u8]) -> Error {
-	let json = serde_json::from_slice::<serde_json::Value>(raw);
+struct ReadTaskParams<R: TransportReceiverT> {
+	receiver: R,
+	close_tx: mpsc::Sender<Result<(), Error>>,
+	to_send_task: mpsc::Sender<FrontToBack>,
+	manager: ThreadSafeRequestManager,
+	max_buffer_capacity_per_subscription: usize,
+}
 
-	let json_str = match json {
-		Ok(json) => serde_json::to_string(&json).expect("valid JSON; qed"),
-		Err(e) => e.to_string(),
+async fn read_task<R>(params: ReadTaskParams<R>)
+where
+	R: TransportReceiverT,
+{
+	let ReadTaskParams { receiver, close_tx, to_send_task, manager, max_buffer_capacity_per_subscription } = params;
+
+	let backend_event = futures_util::stream::unfold(receiver, |mut receiver| async {
+		let res = receiver.receive().await;
+		Some((res, receiver))
+	});
+
+	let closed = close_tx.closed();
+	let pending_unsubscribes = MaybePendingFutures::new();
```

Comment:
I do wonder, is this only used for `unsubscribes`? Wouldn't the `receiver: R` also provide pings and plain messages?

Reply:
No, this is really an edge-case if a subscription fails (such as request ID is already occupied) or that a subscription can't keep it with server all other things are just handled independently of the send and receive tasks by modifying the RequestManager.  

Thus, this really just is a matter to propagate it to the frontend properly.